src/DynamoCoreWpf/Rendering/HelixRenderPackage.cs

```diff
@@ -796,13 +819,147 @@ public void AddTextureMapForMeshVerticesRange(int startIndex, int endIndex, byte
+        }
+
+        /// <summary>
+        /// Set the transform using a series of doubles. The resulting transform is applied to all geometry in the renderPackage.
+        /// Following conventional matrix notation, m11 is the value of the first row and first column, and m12
+        /// is the value of the first row and second column.
+        /// NOTE: This method should set the matrix exactly as described by the caller.
+        /// </summary>
+        /// <param name="m11"></param>
+        /// <param name="m12"></param>
+        /// <param name="m13"></param>
+        /// <param name="m14"></param>
+        /// <param name="m21"></param>
+        /// <param name="m22"></param>
+        /// <param name="m23"></param>
+        /// <param name="m24"></param>
+        /// <param name="m31"></param>
+        /// <param name="m32"></param>
+        /// <param name="m33"></param>
+        /// <param name="m34"></param>
+        /// <param name="m41"></param>
+        /// <param name="m42"></param>
+        /// <param name="m43"></param>
+        /// <param name="m44"></param>
+        public void AddInstanceMatrix(float m11, float m12, float m13, float m14,
+            float m21, float m22, float m23, float m24,
+            float m31, float m32, float m33, float m34,
+            float m41, float m42, float m43, float m44, Guid id)
+        {
+
+            if (!instanceMeshVerticesRange.ContainsKey(id) && !instanceLineVerticesRange.ContainsKey(id))
+            {
+                throw new Exception("The reference to the mesh range for this ID does not exists.");
+            }
+
+            var instance = new Matrix(m11, m12, m13, m14, m21, m22, m23, m24, m31, m32, m33, m34, m41, m42, m43, m44);
+
+            List<Matrix> items;
+            if (instances.TryGetValue(id, out items))
+            {
+                items.Add(instance);
+            }
+            else
+            {
+                instances.Add(id, new List<Matrix> {instance});
+            }
+        }
+
+        /// <summary>
+        /// Set the transform as a double array, this transform is applied to all geometry in the renderPackage.
+        /// This matrix should be laid out as follows in row vector order:
+        /// [Xx,Xy,Xz, 0,
+        ///  Yx, Yy, Yz, 0,
+        ///  Zx, Zy, Zz, 0,
+        ///  offsetX, offsetY, offsetZ, W]
+        /// NOTE: This method should transform the matrix from row vector order to whatever form is needed by the implementation.
+        /// When converting from ProtoGeometry CoordinateSystem form to input matrix, set the first row to the X axis of the CS,
+        /// the second row to the Y axis of the CS, the third row to the Z axis of the CS, and the last row to the CS origin, where W = 1. 
+        /// </summary>
+        /// <param name="matrix"></param>
+        public void AddInstanceMatrix(float[] matrix, Guid id) 
+        {
+            if (!instanceMeshVerticesRange.ContainsKey(id))
```

Comment:
Do you need to check for `instanceLineVerticesRange` here as well?

Reply:
yes, check line as well.